### ISSUE TITLE
Partial hunk staging as operator

### ISSUE THREAD

**Issue by pedro757** [lewis6991/gitsigns.nvim#421]
It would be cool take advantage of the partial hunk staging feature just implemented making a operator to work with motions like:

`<leader>hs3j --> stage next 3 lines` `<leader>hsip --> stage paragraph` `<leader>hsiw --> stage word` and so on..

**Comment by molleweide** [comment]
Wonderful! I was just about to ask if this was possible to do!

**Comment by pedro757** [comment]
It will be a very unique feature, I've never seen it in other Editors. @molleweide

**Comment by molleweide** [comment]
Haha yeah I think it could be quite interesting feature

**Comment by molleweide** [comment]
Very nice when hunks too big

**Comment by tsyroid** [comment]
All in. Was just searching the thread in advance of posting the following issue:
In `gitgutter` I could work with partial hunks (from the documentation):
```
To stage part of any hunk:

preview the hunk, e.g. <Leader>hp;
move to the preview window, e.g. :wincmd P;
delete the lines you do not want to stage;
stage the remaining lines: either write (:w) the window or stage via <Leader>hs or :GitGutterStageHunk.
```
Really, really would be wonderful if gitsigns would allow me to do the same.
/tms

**Comment by j-xella** [comment]
> All in. Was just searching the thread in advance of posting the following issue: In `gitgutter` I could work with partial hunks (from the documentation):
> 
> ```
> To stage part of any hunk:
> 
> preview the hunk, e.g. <Leader>hp;
> move to the preview window, e.g. :wincmd P;
> delete the lines you do not want to stage;
> stage the remaining lines: either write (:w) the window or stage via <Leader>hs or :GitGutterStageHunk.
> ```
> 
> Really, really would be wonderful if gitsigns would allow me to do the same. /tms

Agree, wanted to ask about it as well. But this should probably be moved to a seprarate issue, no?

**Comment by afonsocarlos** [comment]
+1

That's actually the only thing that's missing from vmi-gitgutter for this plugin to be 10/10

**Comment by lewis6991** [comment]
How does vim-gitgutter implement this, last time I checked it doesn't have *true* partial hunk staging like Gitsigns does, so an operator mapping doesn't make sense.

**Comment by lewis6991** [comment]
btw, not sure if anyone knows this, but if you edit a buffer created by `:Gitsigns diffthis` you can save/write it to stage changes. This is opposed to vim-gitgutter that allows you to edit the hunk preview window.

**Comment by afonsocarlos** [comment]
> btw, not sure if anyone knows this, but if you edit a buffer created by `:Gitsigns diffthis` you can save/write it to stage changes. This is opposed to vim-gitgutter that allows you to edit the hunk preview window.

This is exactly what I, and probably most other users, were looking for. I thought the quick preview window edit very handy, but already having the option to edit hunk before staging is already a game changer for me!
Also, I might've missed it, but I think it would be great to have this feature documented in README

**Comment by afonsocarlos** [comment]
> btw, not sure if anyone knows this, but if you edit a buffer created by `:Gitsigns diffthis` you can save/write it to stage changes. This is opposed to vim-gitgutter that allows you to edit the hunk preview window.

This is exactly what I, and probably most other users, were looking for. I thought the quick preview window edit very handy, but already having the option to edit hunk before staging is already a game changer for me!
Also, I might've missed it, but I think it would be great to have this feature documented in README

**Comment by lewis6991** [comment]
Fwiw fugitive allows you to do that too.

**Comment by xuyangy** [comment]
Visual select the changed lines that you want to stage and run `Gitsigns stage_hunk` (equivalently `:'<,'>Gitsings stage_hunk`).

**Comment by rtalexk** [comment]
I have the following mapping on my config

```lua
map({ 'n', 'v' }, '<leader>ghs', ':Gitsigns stage_hunk<CR>', 'Stage hunk')
```

This allow my to select lines (`V`) then stage selected lines, i.e: `V3j<Space>ghs` translates to _Select the current line, move 3 lines below (4 lines selected) and stage them_.

**Comment by jakubbortlik** [comment]
For other operator mappings, I'm using a pair of helper functions defined in my `~/.config/nvim/lua/utils.lua` file:
```lua
M._operator_callback = nil

function M.operator(mode)
  ---If `mode` is nil, we're running this function from a mapping
  ---Otherwise we're running this from g@{motion} and `mode` is "line", "char", or
  ---"block"
  if mode == nil then
    vim.o.operatorfunc = "v:lua.require'utils'.operator"
    return "g@" -- call the `operatorfunc`, see :h g@
  end
  if M._operator_callback then
    M._operator_callback()
  end
  return "" -- do nothing, the _operator_callback already did the work
end

---Return a function that can be used as RHS in mappings. The mapping must have
---opts = { expr = true }.
---@param callback function The function to execute
M.make_operator = function(callback)
  return function()
    M._operator_callback = callback
    return M.operator()
  end
end
```
Then I can simply create an operator mapping for Gitsigns like this:
```lua
local utils = require("utils")
vim.keymap.set("n", "<leader>H", utils.make_operator(function()
  vim.cmd(vim.fn.line("'[") .. "," .. vim.fn.line("']") .. "Gitsigns stage_hunk")
end), {desc = "Stage hunk(s) the operator moves over", expr = true})
```
